rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+Not only does this seem complex, but also hard to be made efficient.
+
+Another possibility is to have everything evaluate in only one world, either Nix
+or Nickel. Practically, we could compile the Nix code to Nickel or vice-versa,
+and run everything on one side only.
+
+Compiling Nickel to Nix and run everything on the Nix side would have the
+advantage of leveraging all the Nix primitives and machinery freely. However,
+this solution simply nullifies a lot of the value proposition of Nickel. The
+added value of contracts is mainly due to its careful design as a primitive
+feature of the language, and while it could be simulated to some extent in Nix,
+the error message and performances just couldn't be in par. Similarly, the merge
+system being native is in part motivated by the perspective of better
+performance and error message. This simply becomes moot if we compile those
+features away back to plain Nix code.
+
+On the other hand, compiling Nix to Nickel looks more appealing. Nickel is close
+to being a superset of Nix (the only non-trivial missing features being the
+`with` construct and string contexts, but the latter doesn't even really change
+the semantics). This would preserve the Nickel advances, and we could leverage
+directly any non-builtin Nix function, including all the Nixpkgs library
+functions.
+
+The potential drawbacks of this approach:
+
+- Performance: Nickel is not expected to be in par with Nix at this point, so
+  the performance of Nixpkgs compiled to Nickel might be prohibitively bad at
+  first. On a positive note, this would provide a great benchmark suite and
+  motivation to guide future performance improvements.
+- Some builtin may incur effects, which are not trivial to combine with the
+  execution model (revertible thunks). That being said, those effects are in
+  practice idempotent and commutative (creating paths in the store, creating
+  derivation, etc.), which should be fine.
+
+Compilation could be done on the fly, as the compilation process would be rather
+straightforward, or have a Nixpkgs snapshot pre-compiled to Nickel, or a mix of
+both.
+
+This solution requires to reimplement Nix builtins in Nickel (the nickel-nix
+compatibility layer) . This is an interesting milestone in itself, because even
+without a Nix-to-Nickel compiler, the compatibility layer would already make
+writing derivations in pure Nickel possible.
+
+- TODO: what about dynamic imports?
+- TODO: what about using a Nickel pkgs from Nix?
+
+#### Using Nixpkgs in the PARM
+
+Using a package from Nixpkgs in the PARM with a Nix-to-Nickel compiler would be
+straightforward, as we could evaluate anything to a derivation whenever needed.
+One would need to use Nixpkgs functions and idioms to do e.g. overrides though,
+but this seems pretty hard to avoid, as a systematic translation from the
+Nixpkgs model to the PARM doesn't seem trivial, if even doable.
```

Comment:
I don’t think that’s a bad thing in the long term tbh. Having an interface that would be too transparent could cause the path of least resistance to lead to a weird spaghetti of Nix and Nickel code intertangled, which would probably be a very bad thing.

At least, having some annoying constraints at the boundaries (even if just because both worlds have different idioms) mean that people will care about these boundaries, and things will stay clearly separated